### PR TITLE
Added tests for Group service

### PR DESCRIPTION
**Pull request by zanpizmoht**
Added tests for adding, deleting and editing groups, and adding devices to groups.

Signed-off-by: zanpizmoht <zan.pizmoht@comtrade.com>

Test scenarios for Groups are added in this PR.

**Related Issue**
_None_

**Description of the solution adopted**
Test scenarios that are added in this PR are checking functionalities regarding Groups. All scenarios are added to `GroupService.feature`. Scenarios are testing if Groups are correctly created, edited or deleted and if Devices can be added to Groups.

**Screenshots**
/

### CODE REVIEW

**Review by Coduz — APPROVED** [review]

**Comment by Coduz** [comment]
CQ submitted: [CQ 21395](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21395)

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 1]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+      | type    | name                   | value | scopeId | parentId |
+      | boolean | infiniteChildEntities  | true  | 0       | 1        |
+      | integer | maxNumberChildEntities | 5     | 0       | 1        |
+    And I configure the device registry service
+      | type    | name                   | value |
+      | boolean | infiniteChildEntities  | true  |
+      | integer | maxNumberChildEntities | 10    |
+    When I create the group with name "Group1" and description "Description 1234"
+    And I create a device with name "Device1"
+    And I add device to group
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change step name to: "And I add device "Device1" to group "Group1"

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 2]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+    And I search for a group named "Group1"
+    Then Device is in groups Assigned Devices
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change step name to: "Then Device "Device1" is in Assigned Devices of group "Group1" so we can use this more clearly in other scenarios as well.

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 3]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+    And No exception was thrown
+    And I logout
+
+  Scenario: Change Group With Description of Device
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change title to: "Moving Device From One Group With Description to Another

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 4]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+  Login as kapua-sys, go to groups, create a group.
+  Go to devices, create a device ("Device1") and add created group to it.
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change description to: "Go to devices, create a device ("Device1") and add it to the created group."

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 5]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+  Check if created device is in Asigned Devices of the group.
+  Create another group and add it to "Device1"
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change description to: "Create another group and add device ("Device1") to it."

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 6]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+  Check both groups Assigned Devices ("Device1" should only be in second group)
+  Kapua should throw Exception.
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change description to: "Kapua should throw Exception  when checking for device in "Group1"".

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 7]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+      | type    | name                   | value | scopeId | parentId |
+      | boolean | infiniteChildEntities  | true  | 0       | 1        |
+      | integer | maxNumberChildEntities | 5     | 0       | 1        |
+    And I configure the device registry service
+      | type    | name                   | value |
+      | boolean | infiniteChildEntities  | true  |
+      | integer | maxNumberChildEntities | 10    |
+    When I create the group with name "Group1" and description "Description 123"
+    And I create a device with name "Device1"
+    And I add device to group
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change step name to: "And I add device "Device1" to group "Group1""

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 8]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+    When I search for a group named "Group1"
+    Then Device is in groups Assigned Devices
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change step name to: "Then Device "Device1" is in Assigned Devices of group "Group1"

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 9]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+    When I create the group with name "Group2" and description "Description 123"
+    And I change devices group
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change step name to: "And I move device "Device1" from group "Group1" to group "Group2"

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 10]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+    And I search for a group named "Group2"
+    Then Device is in groups Assigned Devices
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change step name to: "Then Device "Device1" is in Assigned Devices of group "Group2"

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 11]
```diff
@@ -256,3 +256,1053 @@ Feature: Group Service tests
...
+    When I search for a group named "Group1"
+    Then Device is not in groups Assigned Devices
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Change step name to: "Then Device "Device1" is not in Assigned Devices of group "Group1""

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 12]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
 
         assertEquals(device.getClientId(), deviceName);
     }
+
+    @And("^I add device to group$")
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Change step name to: "I add device "Device1" to group "Group1""

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 13]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+    public void iAddDeviceToGroup() throws Exception{
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Change method name to: "public void iAddDeviceToGroup(String deviceName, String groupName) throws Exception{

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 14]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+        Device device = (Device) stepData.get("Device");
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Change method implementation, so it accepts the parameters that are given to it; do not use "stepData.get("Device")

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 15]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+        Group group = (Group) stepData.get("Group");
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Change method implementation, so it accepts the parameters that are given to it; do not use "stepData.get("Group")

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 16]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+            KapuaId groupId = group.getId();
+            device.setGroupId(groupId);
+            Device newDevice = deviceRegistryService.update(device);
+            stepData.put("Device", newDevice);
+        } catch (Exception e) {
+            verifyException(e);
+        }
+    }
+
+    @When("^I change devices group to No Group$")
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Change step name to: "Then Device "Device1" is removed from all groups".

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 17]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+    public void iChangeDevicesGroupToNoGroup() throws Exception{
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Add parameter to the method: "public void iChangeDevicesGroupToNoGroup(String deviceName) throws Exception{"

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 18]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+        Device device = (Device) stepData.get("Device");
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Use "Device1" paramater and not "last" device.

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 19]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+        try {
+            device.setGroupId(null);
+            device = deviceRegistryService.update(device);
+            stepData.put("Device", device);
+        } catch (Exception e) {
+            verifyException(e);
+        }
+    }
+
+    @And("^I create (\\d+) devices and add them to group$")
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Change step name to: "@And("^I create (\\d+) devices and add them to group "Group1"$")

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 20]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+    public void iCreateDevicesAndAddThemToGroup(int numberOfDevices) throws Exception {
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: Change method implementation so it accepts group parameter: " public void iCreateDevicesAndAddThemToGroup(int numberOfDevices, String group) throws Exception {"

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 21]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+            for(int i = 0; i < numberOfDevices; i++) {
+                iCreateADeviceWithName(String.format("Device%02d", i));
+                iAddDeviceToGroup();
+            }
+        } catch (Exception e) {
+            verifyException(e);
+        }
+    }
+
+    @And("^Device has no group$")
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: This is not used anywhere?

File: service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java [hunk 22]
```diff
@@ -2559,4 +2560,65 @@ public void iFindDeviceWithClientId(String deviceName) {
...
+    public void deviceHasNoGroup() throws Exception {
+        Device device = (Device) stepData.get("Device");
+        try {
+            assertNull(device.getGroupId());
+        } catch (Exception e) {
+            verifyException(e);
+        }
+    }
+
+    @And("^I update device$")
```
Comment (LeoNerdoG) [inline service/device/registry/test-steps/src/main/java/org/eclipse/kapua/service/device/registry/steps/DeviceRegistrySteps.java]: This is also not used here?

**Review by LeoNerdoG — CHANGES_REQUESTED** [review]
Hi @zanpizmoht,
I have reviewed the PR and there are some changes that need to be made.

**Review by LeoNerdoG — CHANGES_REQUESTED** [review]
Hi @zanpizmoht,
I have reviewed the PR and there are some changes that need to be made.

File: qa/integration/src/test/resources/features/authorization/GroupService.feature [hunk 23]
```diff
@@ -256,3 +256,792 @@ Feature: Group Service tests
...
+    Given I login as user with name "kapua-sys" and password "kapua-password"
+    And I configure the group service
+      | type    | name                   | value | scopeId | parentId |
+      | boolean | infiniteChildEntities  | true  | 0       | 1        |
+      | integer | maxNumberChildEntities | 5     | 0       | 1        |
+    When I create the group with name "Group1" and description "a"
+    Then No exception was thrown
+    Then I logout
+
+  Scenario: Creating Unique Tag With Long Description
```
Comment (LeoNerdoG) [inline qa/integration/src/test/resources/features/authorization/GroupService.feature]: Please change this title to: "Creating Unique Group With Long Description" becasue we are creating groups here and not tags =)

**Review by LeoNerdoG — COMMENTED** [review]
Hey, @zanpizmoht, there are only 2 minor things left to change. 
Please change them and push to PR. 
@Coduz this PR can be labeled as "QA Approved" label.

**Review by LeoNerdoG — APPROVED** [review]
Hi @zanpizmoht,
I have checked all the changes and everything is OK now. 
@coduz, can you please restart the failing tests and merge when CQ passes?

**Review by LeoNerdoG — APPROVED** [review]

**Comment by Coduz** [comment]
[CQ Approved!](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21395#c6) 🎉

File: qa/integration/src/test/resources/features/user/UserPermissionI9n.feature [hunk 24]
```diff
@@ -1441,8 +1441,13 @@ Feature: User Permission tests
       | account | delete |
     And A device named "test_device"
     And I create a job with the name "test_job"
+<<<<<<< HEAD
     And I create tag with name "test_tag" without description
     And I create the group with name "test_group"
+=======
+    And I create a tag with name "test_tag"
+    And I create a group with name "test_group"
+>>>>>>> Added tests for Group service
```
Comment (Coduz) [inline qa/integration/src/test/resources/features/user/UserPermissionI9n.feature]: Please remove the merge meta data

**Review by Coduz — CHANGES_REQUESTED** [review]

**Comment by zanpizmoht** [comment]
Hi @Coduz,
I fixed the errors and I saw Travis just passed so this PR is ready to be merged.
Regards, Zan

**Review by Coduz — APPROVED** [review]

**Pull request merged**